tests/deploy/ui/test_ui.py

```diff
@@ -0,0 +1,183 @@
+            api_app(
+                config=self.config,
+                ignore_unavailable_components=True,
+            ),
+            host=self.config.api.hostname,
+            port=self.config.api.port,
+        )
+
+    def server_up(self):
+        try:
+            return httpx.get(self.config.api.url).is_success
+        except httpx.ConnectError:
+            return False
+
+    def start(self):
+        self.proc = Process(target=self.start_server, args=(), daemon=True)
+        self.proc.start()
+
+        timeout = 5
+        while timeout > 0 and not self.server_up():
+            print(f"Waiting for API server to come up on {self.config.api.url}")
+            time.sleep(1)
+            timeout -= 1
+
+    def stop(self):
+        self.proc.kill()
+
+
+@pytest.fixture
+def api_server(config):
+    server = ApiServer(config)
+    try:
+        server.start()
+        yield server
+    finally:
+        server.stop()
+
+
+def auth_header(base_url):
+    username = default_user()
+    token = (
+        httpx.post(
+            base_url + "/token",
+            data={
+                "username": username,
+                "password": os.environ.get(
+                    "RAGNA_DEMO_AUTHENTICATION_PASSWORD", username
+                ),
+            },
+        )
+        .raise_for_status()
+        .json()
+    )
+
+    return f"Bearer {token}"
+
+
+@pytest.fixture(scope="function")
+def page(config, api_server, headed_mode):
+    server = ui_app(config=config, open_browser=False)
+
+    with sync_playwright() as playwright:
+        server.serve()
+        browser = playwright.chromium.launch(headless=not headed_mode)
+        context = browser.new_context()
+        page = context.new_page()
+
+        yield page
+
+        context.close()
+        browser.close()
+        pn.state.kill_all_servers()
+
+
+def test_health(config, page) -> None:
+    health_url = config.ui.origins[0] + "/health"
+    page.goto(health_url)
+    expect(page.get_by_role("heading", name="Ok")).to_be_visible()
+
+
+def test_index_with_blank_credentials(config, page) -> None:
+    # Index page, no auth
+    index_url = config.ui.origins[0]
+    page.goto(index_url)
+    expect(page.get_by_role("button", name="Sign In")).to_be_visible()
+
+    # Authorize with no credentials
+    page.get_by_role("button", name="Sign In").click()
+
+    expect(page.get_by_role("button", name=" New Chat")).to_be_visible()
+
+
+@pytest.mark.skip(reason="Need to figure out how to set the auth token")
+def test_index_with_auth_token(config, page) -> None:
+    auth = auth_header(config.api.url)
+    assert len(auth) > len("Bearer ")
+
+    page.set_extra_http_headers({"Authorization": auth})  # this doesn't work
+    pn.state.cookies["auth_token"] = ""  # or this
+    pn.state.headers["Authorization"] = auth  # or this
+
+    index_url = config.ui.origins[0]
+    page.goto(index_url)
+    expect(page.get_by_role("button", name=" New Chat")).to_be_visible()
```

Comment:
I don't understand this test. Why are you trying to do here? The `auth_token` is a cookie that has to be set by the UI. What you can and should test is that after the login you now have the respective cookie set.

Ultimately you can use this cookie to access the API with bearer authorization and check if the information in the UI is correct. For example, when getting an answer, check with the API whether the displayed sources are actually the ones we also have in the DB.

Reply:
Oh I see, I was trying to test if having the auth token already in the browser would bypass the login page and route directly to the main index.

Checking whether the displayed sources match the DB is a great test idea and I will add that.